haystack/core/pipeline/base.py

```diff
@@ -816,6 +816,107 @@ def _init_graph(self):
+        to_run: List[Tuple[str, Component]],
+        waiting_for_input: List[Tuple[str, Component]],
+    ):
+        """
+        Distributes the output of a Component to the next Components that need it.
+
+        This also updates the queues that keep track of which Components are ready to run and which are waiting for input.
+
+        :param component_name: Name of the Component that created the output
+        :param component_result: The output of the Component
+        :paramt inputs_by_component: The current state of the inputs divided by Component name
+        :param to_run: Queue of Components to run
+        :param waiting_for_input: Queue of Components waiting for input
+
+        :return: The updated output of the Component without the keys that were distributed to other Components
+        """
+        # We keep track of which keys to remove from res at the end of the loop.
+        # This is done after the output has been distributed to the next components, so that
+        # we're sure all components that need this output have received it.
+        to_remove_from_res = set()
+
+        for _, receiver_name, connection in self.graph.edges(nbunch=component_name, data=True):
+            sender_socket: OutputSocket = connection["from_socket"]
+            receiver_socket: InputSocket = connection["to_socket"]
+
+            if sender_socket.name not in component_result:
+                # This output wasn't created by the sender, nothing we can do
+                continue
+
+            if receiver_name not in inputs_by_component:
+                inputs_by_component[receiver_name] = {}
+
+            # We'll remove this key from the output at the end of the loop
+            to_remove_from_res.add(sender_socket.name)
+
+            value = component_result[sender_socket.name]
+
+            if receiver_socket.is_variadic:
+                # Variadic inputs are always lists
+                if receiver_socket.name not in inputs_by_component[receiver_name]:
+                    # Create the list if it doesn't exist
+                    inputs_by_component[receiver_name][receiver_socket.name] = []
+                inputs_by_component[receiver_name][receiver_socket.name].append(value)
+            else:
+                inputs_by_component[receiver_name][receiver_socket.name] = value
+
+            receiver = self.graph.nodes[receiver_name]["instance"]
+            pair = (receiver_name, receiver)
+
+            is_greedy = getattr(receiver, "__haystack_is_greedy__", False)
+            if receiver_socket.is_variadic and is_greedy:
+                # If the receiver is greedy, we can run it right away.
+                # First we remove it from the status lists it's in if it's there or we risk running it multiple times.
+                if pair in to_run:
+                    to_run.remove(pair)
+                if pair in waiting_for_input:
+                    waiting_for_input.remove(pair)
+                to_run.append(pair)
```

Comment:
Why does the component need to be moved to the back of the queue in this case?

Reply:
All Components are appended to `to_run`.

@masci and I decided to add this concept of greediness when we reworked the run logic to differentiate between two different type of Components with `Variadic` input. Those that must run as soon as they have an input, and those that must wait as long as possible before running to receive as many inputs as possible.

We noticed that in the previous implementation `Variadic`s some times were assumed to run as soon as they got input and some other times not. So we decided to formalize it.

As of now the only Component that is greedy and `Variadic` is the `BranchJoiner`, the old `Multiplexer`.

Reply:
Then the component should be move to the head of the queue, not the tail.

Reply:
After discussing it internally, we've decided to keep the logic as-is for now and investigate it more thoroughly once the refactoring is complete.